xdcs-agent/src/main/python/xdcs/kernel.py

```diff
@@ -0,0 +1,211 @@
+    _ctx: cl.Context
+
+    def __init__(self, kernel_config: OpenClKernelConfig):
+        self._kernel_config = kernel_config
+        self._host_argument_values = []
+        self._device_argument_values = []
+
+    @staticmethod
+    def from_config(grpc_config: KernelConfig, kernel_params: [dict]):
+        return KernelManager(OpenClKernelConfig.parse_config(grpc_config, kernel_params))
+
+    @property
+    def kernel_config(self):
+        return self._kernel_config
+
+    @property
+    def device_argument_values(self):
+        return self._device_argument_values
+
+    def prepare_execution(self, agent_variables: dict):
+        self.fetch_argument_values()
+        self.set_agent_partitioning_variables(agent_variables)
+        self.initialize_context()
+        self.prepare_argument_values()
+
+    def execute_kernel(self, kernel_source: str, kernel_name: str):
+        self._queue = Compiler.OpenClKernelRunner.run(
+            kernel_source,
+            kernel_name,
+            self._ctx,
+            self.kernel_config.global_work_shape,
+            self.kernel_config.local_work_shape,
+            self.device_argument_values
+        )
+
+    def initialize_context(self):
+        # TODO based on selected resources during task creation
+        self._ctx = cl.create_some_context()
+
+    def set_agent_partitioning_variables(self, agent_variables: dict):
+        for argument in self._kernel_config.arguments:
+            if argument.name in OpenClKernelConfig.reserved_arg_names:
+                if argument.type != KernelArgument.Type.SIMPLE or argument.direction != KernelArgument.Direction.IN:
+                    raise Exception('Wrong argument config with reserved name: %s' % str(argument))
+                argument.value = agent_variables[argument.name]
+
+    def fetch_argument_values(self):
+        object_ids: [str] = []
+        for argument in self._kernel_config.arguments:
+            object_ids.append(argument.value)
+        # remove duplicates due to bug in Retrieving objects
+        object_ids = list(dict.fromkeys(object_ids))
+        xdcs().execute(RetrieveObjectsCmd((ids for ids in [object_ids])))
+        for argument in self._kernel_config.arguments:
+            value: str = xdcs().object_repository().cat_bytes(argument.value).decode('utf-8')
+            argument.value = value
+
+    def prepare_argument_values(self):
+        mf = cl.mem_flags
+        for argument in self._kernel_config.arguments:
+            if argument.type == KernelArgument.Type.SIMPLE:
+                value = KernelManager.interpret_simple_value(argument.value)
+                self._host_argument_values.append(value)
+                self._device_argument_values.append(value)
+            elif argument.type == KernelArgument.Type.POINTER:
+                if argument.direction == KernelArgument.Direction.IN:
+                    host_buffer = np.frombuffer(buffer=argument.value.encode('utf-8'), dtype=np.int8)
```

Comment:
value nie powinno być czytane z OR?

Reply:
było czytane, ale do przechowywania referencji z OR oraz wartości było używane to samo pole, zmieniłem to rozwiązanie na czytelniejsze: teraz mamy 2 pola: obj_id oraz value. Na początku value jest niezdefiniowane, ale mamy wszystki obj_id. Następnie robimy fetch z OR wszystkich objektów z id = obj_id w naszych kernel argumentach i wyciągnięte wartości przypisujemy do value, robimy później dalszy processing value jeżeli potrzeba (na przykład nadpisujemy argument.value w sytuacji, gdy ma zarezerwowaną nazwę: XDCS_AGENT_ID lub XDCS_AGENT_COUNT)